http/server.ts

```diff
@@ -313,65 +289,61 @@ async function readRequest(
   return [req, err];
 }
 
-function maybeHandleReq(
-  env: ServeEnv,
-  conn: Conn,
-  maybeReq: [ServerRequest, BufState]
-): void {
-  const [req, _err] = maybeReq;
-  if (_err) {
-    conn.close(); // assume EOF for now...
-    return;
+export class Server implements AsyncIterableIterator<ServerRequest> {
+  private closing = false;
+  private looping = false;
+  private queue: AsyncQueue<ServerRequest> = new AsyncQueue();
+
+  constructor(public listener: Listener) {}
+
+  async acceptLoop(): Promise<void> {
+    assert(!this.looping);
+    this.looping = true;
+    try {
+      while (!this.closing) {
+        const conn = await this.listener.accept();
+        this.serveConn(conn);
+      }
+    } finally {
+      this.looping = false;
+    }
   }
-  env.reqQueue.push(req); // push req to queue
-  env.serveDeferred.resolve(); // signal while loop to process it
-}
 
-function serveConn(env: ServeEnv, conn: HttpConn, bufr?: BufReader): void {
-  readRequest(conn, bufr).then(maybeHandleReq.bind(null, env, conn));
-}
+  close(): void {
+    this.listener.close();
+    this.closing = true;
+  }
 
-export async function* serve(
-  addr: string
-): AsyncIterableIterator<ServerRequest> {
-  const listener = listen("tcp", addr);
-  const env: ServeEnv = {
-    reqQueue: [], // in case multiple promises are ready
-    serveDeferred: deferred()
-  };
-
-  // Routine that keeps calling accept
-  let handleConn = (_conn: Conn): void => {};
-  let scheduleAccept = (): void => {};
-  const acceptRoutine = (): void => {
-    scheduleAccept = (): void => {
-      listener.accept().then(handleConn);
-    };
-    handleConn = (conn: Conn): void => {
-      const httpConn = createHttpConn(conn);
-      serveConn(env, httpConn); // don't block
-      scheduleAccept(); // schedule next accept
-    };
-
-    scheduleAccept();
-  };
-
-  acceptRoutine();
-
-  // Loop hack to allow yield (yield won't work in callbacks)
-  while (true) {
-    await env.serveDeferred.promise;
-    env.serveDeferred = deferred(); // use a new deferred
-    let queueToProcess = env.reqQueue;
-    env.reqQueue = [];
-    for (const result of queueToProcess) {
-      yield result;
-      // Continue read more from conn when user is done with the current req
-      // Moving this here makes it easier to manage
-      serveConn(env, result.conn, result.r);
+  async serveConn(conn: Conn): Promise<void> {
+    const httpConn = createHttpConn(conn);
+    const bufr = new BufReader(httpConn);
+    while (true) {
+      const [req, err] = await readRequest(httpConn, bufr);
+      if (err) {
+        // TODO(ry) This should be more granular. Perhaps return back a 400 or
+        // 500 error?
+        httpConn.close();
+        break;
+      }
+      this.queue.push(req);
```

Comment:
This does not provide proper backpressure semantics. 
You should make sure that readRequest is not getting called when the iterator isn't being iterated.
Basically, the 'queue' thing needs to go away.

Reply:
Done, using your latch code. PTAL.